apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/RpcRetryConsumer.cs

```diff
@@ -0,0 +1,215 @@
+                throw;
+            }
+        }
+
+        /// <inheritdoc />
+        public Task ReceiveAsync(IEnumerable<T> items, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            GaxPreconditions.CheckNotNull(items, nameof(items));
+            try
+            {
+                return _consumer.ReceiveAsync(items, cancellationToken);
+            }
+            catch (RpcException e) when (!HandleException(items, e))
+            {
+                throw;
+            }
+        }
+
+        /// <inheritdoc />
+        public void Dispose() => _timer?.Dispose();
+
+        /// <summary>
+        /// Handles a <see cref="RpcException"/> that occurred during a <see cref="Receive(IEnumerable{T})"/> call.
+        /// </summary>
+        /// <returns>True if the exception was handled, false otherwise.</returns>
+        private bool HandleException(IEnumerable<T> items, RpcException e)
+        {
+            // If we are not retrying failed calls either propagate or swallow the exception.
+            if (_options.RetryType == RetryType.None)
+            {
+                switch (_options.ExceptionHandling)
+                {
+                    case ExceptionHandling.Propagate:
+                        return false;
+                    case ExceptionHandling.Ignore:
+                        Debug.WriteLine($"Swallowed exception: {e}");
+                        return true;
+                    default:
+                        Debug.Fail($"Invalid ExceptionHandling option: {_options.ExceptionHandling}");
+                        return true;
+                }
+            }
+            
+            // Get the size of items to be buffered and retried.
+            int size = items.Sum(item => _sizer(item));
+
+            // If the new items would be greater than the size of the buffer drop them or throw.
+            if (size + _sizeInBytes > _options.BufferSizeBytes) 
+            {
+                switch (_options.BufferOverflow)
+                {
+                    case BufferOverflow.IgnoreNewEntries:
+                        Debug.WriteLine($"Ignored {items.Count()} entries.");
+                        return true;
+                    case BufferOverflow.ThrowException:
+                        throw new InvalidOperationException("Buffer overflow", e);
+                    default:
+                        Debug.Fail($"Invalid BufferOverflow option: {_options.BufferOverflow}");
+                        return true;
+                }
+            }
+
+            // Lock the semaphore and add the new items to the buffer.
+            _semaphore.Wait();
+            _sizeInBytes += size;
+            _buffer.AddRange(items);
+            _semaphore.Release();
+            return true;
+        }
+
+        /// <summary>
+        /// Try to flush the buffer if it has any entries.
+        /// </summary>
+        private void FlushBuffer()
+        {
+            _semaphore.Wait();
```

Comment:
As mentioned above, this needs to be released when you return below. I would suggest using a lock object with double-checked locking here:

```csharp
if (_buffer.Count() == 0)
{
  return;
}

List<T> toSend;
int toSendSize;
lock (_lockObject)
{
  if (_buffer.Count() == 0)
  {
    return;
  }
  toSend = _buffer;
  toSendSize = _sizeInBytes;
  _buffer = new List<T>();
  _retries++;
}
...
```